Source/Scene/parseKTX2.js

```diff
@@ -0,0 +1,241 @@
+import arraySlice from "../Core/arraySlice.js";
+import defined from "../Core/defined.js";
+import PixelFormat from "../Core/PixelFormat.js";
+import RuntimeError from "../Core/RuntimeError.js";
+import VulkanConstants from "../Core/VulkanConstants.js";
+import WebGLConstants from "../Core/WebGLConstants.js";
+import { read } from "../ThirdParty/Workers/ktx-parse.modern.js";
+
+var fileIdentifier = [
+  0xab, // '«'
+  0x4b, // 'K'
+  0x54, // 'T'
+  0x58, // 'X'
+  0x20, // ' '
+  0x32, // '2'
+  0x30, // '0'
+  0xbb, // '»'
+  0x0d, // '\r'
+  0x0a, // '\n'
+  0x1a, // '\x1A'
+  0x0a, // '\n'
+];
+
+var faceOrder = [
+  "positiveX",
+  "negativeX",
+  "positiveY",
+  "negativeY",
+  "positiveZ",
+  "negativeZ",
+];
+
+// Flags
+var colorModelETC1S = 163;
+var colorModelUASTC = 166;
+
+/**
+ * Parses and transcodes KTX2 buffers to an appropriate target format.
+ *
+ * @param {Uint8Array} data Data representing one KTX2 texture.
+ * @param {Object} supportedTargetFormats Target formats available on the current system.
+ * @param {Object} transcoderModule basis_transcoder object.
+ *
+ * @private
+ */
+function parseKTX2(data, supportedTargetFormats, transcoderModule) {
+  var byteBuffer = new Uint8Array(data);
+
+  if (!defined(supportedTargetFormats)) {
+    throw new RuntimeError(
+      "KTX2 Loader does not know which GPU formats are supported."
+    );
+  }
+
+  var isKTX2 = true;
+  var i;
+  for (i = 0; i < fileIdentifier.length; ++i) {
+    if (fileIdentifier[i] !== byteBuffer[i]) {
+      isKTX2 = false;
+      break;
+    }
+  }
+
+  if (!isKTX2) {
+    throw new RuntimeError("Invalid KTX2 file.");
+  }
+
+  // Get Texture Based on Format
+  var header = read(data);
+  var dfd = header.dataFormatDescriptor[0];
+
+  var result = new Array(header.levelCount);
+
+  if (header.layerCount !== 0) {
+    throw new RuntimeError("KTX2 Videos are not supported");
+  }
+
+  if (header.pixelDepth !== 0) {
+    throw new RuntimeError("KTX2 3D textures are unsupported.");
+  }
+
+  if (
+    header.vkFormat === 0x0 &&
+    (dfd.colorModel === colorModelETC1S || dfd.colorModel === colorModelUASTC)
+  ) {
+    // Compressed, initialize transcoder module
+    transcodeCompressed(
+      data,
+      header,
+      supportedTargetFormats,
+      transcoderModule,
+      result
+    );
+  } else {
+    parseUncompressed(header, result);
+  }
+
+  return result;
+}
+
+// Parser for uncompressed
+function parseUncompressed(header, result) {
+  var internalFormat =
+    header.vkFormat === VulkanConstants.VK_FORMAT_R8G8B8A8_SRGB
+      ? PixelFormat.RGBA
+      : PixelFormat.RGB;
+  internalFormat =
+    header.vkFormat === VulkanConstants.VK_FORMAT_B10G11R11_UFLOAT_PACK32
+      ? WebGLConstants.R11F_G11F_B10F
+      : internalFormat;
+  internalFormat =
```

Comment:
@lilleyse Since the math in the cubemap shader has to be different for `..._UNORM` formats, what would be a good way to catch that case here? Should this be an `if (header.vkFormat === VulkanConstants.R8G8B8A8_UNORM || header.vkFormat === VulkanConstants.R16G16B16A16_UNORM ...)` statement?